borsh-derive/src/internals/serialize/structs/mod.rs

```diff
@@ -19,45 +23,92 @@ pub fn process(input: &ItemStruct, cratename: Path) -> syn::Result<TokenStream2>
             for field in &fields.named {
                 let field_id = serialize::FieldId::Struct(field.ident.clone().unwrap());
 
-                process_field(field, field_id, &cratename, &mut generics_output, &mut body)?;
+                process_field::<IS_ASYNC>(
+                    field,
+                    field_id,
+                    &cratename,
+                    &mut generics_output,
+                    &mut body,
+                )?;
             }
         }
         Fields::Unnamed(fields) => {
             for (field_idx, field) in fields.unnamed.iter().enumerate() {
                 let field_id = serialize::FieldId::new_struct_unnamed(field_idx)?;
 
-                process_field(field, field_id, &cratename, &mut generics_output, &mut body)?;
+                process_field::<IS_ASYNC>(
+                    field,
+                    field_id,
+                    &cratename,
+                    &mut generics_output,
+                    &mut body,
+                )?;
             }
         }
         Fields::Unit => {}
     }
-    generics_output.extend(&mut where_clause, &cratename);
+    generics_output.extend::<IS_ASYNC>(&mut where_clause, &cratename);
+
+    let serialize_trait = Ident::new(
+        if IS_ASYNC {
+            "BorshSerializeAsync"
+        } else {
+            "BorshSerialize"
+        },
+        Span::call_site(),
+    );
+    let writer_trait_path = if IS_ASYNC {
+        quote! { async_io::AsyncWrite }
+    } else {
+        quote! { io::Write }
+    };
+    let r#async = IS_ASYNC.then(|| Token![async](Span::call_site()));
+    let lifetime = IS_ASYNC.then(|| Lifetime::new("'async_variant", Span::call_site()));
```

Comment:
suggest changing/shortening this to `__async` in the style of #260, but this change isn't mandatory/important